### ISSUE TITLE
Deduction guide detection doesn't work with clang

### ISSUE THREAD

**Issue by dan131riley** [oneapi-src/oneTBB#284]
The tbb_config.h tests for clang support of deduction guides and invokability are incorrect, and as a result the simple test program from the tbb documentation,
```
   std::vector<int> v = {0, 1, 2};
   tbb::concurrent_vector cv(v.begin(), v.end());
```
fails in type deduction.  The problem is with the lines
https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/include/tbb/tbb_config.h#L237-L238
The clang `__cpp_deduction_guides` and `__cpp_lib_is_invocable` have date-stamp values and are not testable with `__has_feature`; results of this test snippet
```
   std::cout << __cpp_deduction_guides << " " << __has_feature(__cpp_deduction_guides) << std::endl;
   std::cout << __cpp_lib_is_invocable << " " << __has_feature(__cpp_lib_is_invocable) << std::endl;
```
compiled with clang 5.0 prints the values
```
201611 0
201703 0
```
see https://godbolt.org/z/T63bWr

So those tests should be something like
```
    #define __TBB_CPP17_DEDUCTION_GUIDES_PRESENT            (__cpp_deduction_guides >= 201611L)
    #define __TBB_CPP17_INVOKE_RESULT_PRESENT               (__cpp_lib_is_invocable >= 201703L)
```

**Comment by pavelkumbrasev** [comment]
Hi, @dan131riley,
Thank you for the report. The error does exists. Will be fixed in one of the next updates.

**Comment by jesseschalken** [comment]
The incorrect check for `__cpp_lib_is_invocable`, causes `iterators.h` to use `std::result_of` instead of `std::invoke_result` on this line: 

https://github.com/oneapi-src/oneTBB/blob/427c252e0bb9e191767a62d8a744b21950c343f6/include/tbb/iterators.h#L243-L247


But `std::result_of` does not exist in C++20. Therefore TBB cannot be imported in Clang in C++20 mode, which in turn means that `clang-tidy` does not work on any code that includes TBB headers.

**Comment by kboyarinov** [comment]
Hi, @dan131riley,

The issue with incorrect detection of Class Template Argument Deduction feature on Clang was fixed in oneTBB 2021.2.0.
If you still have any issues, feel free to contact us.

@jesseschalken, regarding the question about `iterators.h` - this header file was removed from oneTBB and now it is a part of [oneDPL](https://github.com/oneapi-src/oneDPL).
